plutus-contract/src/Wallet/Emulator/Notify.hs

```diff
@@ -0,0 +1,99 @@
+{-# LANGUAGE DeriveAnyClass     #-}
+{-# LANGUAGE DeriveGeneric      #-}
+{-# LANGUAGE DerivingStrategies #-}
+{-# LANGUAGE FlexibleContexts   #-}
+{-# LANGUAGE GADTs              #-}
+{-# LANGUAGE LambdaCase         #-}
+{-# LANGUAGE NamedFieldPuns     #-}
+{-# LANGUAGE RankNTypes         #-}
+{-# LANGUAGE TemplateHaskell    #-}
+{-# LANGUAGE TypeOperators      #-}
+-- | Implements contract instance notifications in the emulator.
+module Wallet.Emulator.Notify(
+    EmulatorContractNotifyEffect(..)
+    , sendAgentNotification
+    , handleContractRuntime
+    -- * Contract instance IDs
+    , walletInstanceId
+    , instanceIdWallet
+    -- * Logging
+    , EmulatorNotifyLogMsg(..)
+    ) where
+
+import           Control.Monad.Freer
+import           Control.Monad.Freer.Log   (LogMsg, logWarn)
+import           Control.Monad.Freer.TH    (makeEffect)
+import           Data.Aeson                (FromJSON, ToJSON)
+import           Data.Text.Prettyprint.Doc
+import qualified Data.UUID                 as UUID
+import           GHC.Generics              (Generic)
+
+
+import           Wallet.Effects            (ContractRuntimeEffect (..))
+import           Wallet.Emulator.Wallet    (Wallet (..))
+import           Wallet.Types              (ContractInstanceId (..), Notification (..), NotificationError (..))
+
+data EmulatorContractNotifyEffect r where
+    SendAgentNotification :: Wallet -> Wallet -> Notification -> EmulatorContractNotifyEffect (Maybe NotificationError)
+
+makeEffect ''EmulatorContractNotifyEffect
+
+{- Note [Emulator contract instances]
+
+In the emulator, unlike the SCB, there is no distinction between agents (called
```

Comment:
We should probably fix this!

Reply:
Yeah, this is going to be part of https://jira.iohk.io/browse/SCP-737 ...